client/components/ParsedMessage.vue

```diff
@@ -11,12 +19,50 @@ export default defineComponent({
 		message: {type: Object as PropType<ClientMessage | string>, required: false},
 		network: {type: Object as PropType<ClientNetwork>, required: false},
 	},
+	setup() {
+		const store = useStore();
+
+		return {
+			store,
+		};
+	},
 	render(context) {
-		return parse(
+		const parsed = parse(
 			typeof context.text !== "undefined" ? context.text : context.message.text,
 			context.message,
 			context.network
 		);
+
+		if (context?.message?.type === "message" && this.store.state.settings.parseMd) {
+			const htmls: Map<number, ParseFragment> = new Map();
+			const id = idGenerator();
+
+			const generateStandIns = (nodes) => {
+				const result: string[] = [];
+
+				for (let i = 0; i < nodes.length; i++) {
+					if (nodes[i] instanceof Array) {
+						result.push(...generateStandIns(nodes[i]));
+					} else {
+						if (typeof nodes[i] === "string") {
+							result.push(nodes[i]);
+						} else {
+							const nextId = id();
+
+							htmls.set(nextId, nodes[i]);
+							result.push(createPlaceholder(nextId));
+						}
+					}
+				}
+
+				return result;
+			};
+
+			const toParse = "".concat(...generateStandIns(parsed));
+			return rehydrate(parseMd(toParse, this.store.state.settings.renderMdSrc), htmls);
+		}
```

Comment:
I feel this logic is a bit malplaced.

I think it'd be more in the spirit of what this component logic already does (call a single parse function) if you were to break out this logic into its own function which takes "parsed" as an input parameter, as well as other necessary inputs:
````js
// FILE: components/ParsedMessage.vue

import parseMarkdown from "../js/helpers/mdHelpers";
// [...]
{
  // [...]
  render(context) {
    const parsed = parse(
	  typeof context.text !== "undefined" ? context.text : context.message.text,
      context.message,
      context.network
	);
	if (this.store.state.settings.parseMd) {
      return parseMarkdown(parsed, context.message, this.store.state.settings.renderMdSrc);
	}
	return parsed;
  }
  // [..]
}

// FILE: js/helpers/mdHelper
// [...]
export default function parseMarkdown(parsed, message, renderMdSrc) {
  // no-op on non-message type messages
  if (message?.type !== "message") return input;
  // [...] - rest of logic
}
```

It declutters the component render function, the new function is responsible for deciding whether a message should go through the parser or not (by returning early), and the only logic remaining in the component is checking the parseMd flag in order to know if we should call the markdown parser at all.

Reply:
Just saw this, think it got lost in my inbox. Great idea, thanks! I split it off like you suggested.